rclpy/src/rclpy/impl/common.h

```diff
@@ -0,0 +1,125 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef RCLPY__IMPL__COMMON_H_
+#define RCLPY__IMPL__COMMON_H_
+
+#include <Python.h>
+
+#include <rmw/types.h>
+
+typedef void * create_ros_message_signature (void);
+typedef void destroy_ros_message_signature (void *);
+typedef bool convert_from_py_signature (PyObject *, void *);
+typedef PyObject * convert_to_py_signature (void *);
+
+/// Convert a C rmw_qos_profile_t into a Python QoSProfile object
+/**
+ * \param[in] void pointer to a rmw_qos_profile_t structure
+ * \return QoSProfile object
+ */
+static PyObject *
+rclpy_convert_to_py_qos_policy(void * profile)
+{
+  PyObject * pyqos_module = PyImport_ImportModule("rclpy.qos");
+  if (!pyqos_module) {
+    return NULL;
+  }
+
+  PyObject * pyqos_policy_class = PyObject_GetAttrString(pyqos_module, "QoSProfile");
+  if (!pyqos_policy_class) {
+    return NULL;
+  }
+
+  PyObject * pyqos_profile = PyObject_CallObject(pyqos_policy_class, NULL);
+  Py_DECREF(pyqos_policy_class);
+  if (!pyqos_profile) {
+    return NULL;
+  }
+
+  rmw_qos_profile_t * qos_profile = (rmw_qos_profile_t *)profile;
+
+  PyObject * pyqos_depth = PyLong_FromSize_t(qos_profile->depth);
+  if (!pyqos_depth) {
+    Py_DECREF(pyqos_profile);
+    return NULL;
+  }
+
+  PyObject * pyqos_history = PyLong_FromUnsignedLong(qos_profile->history);
+  if (!pyqos_history) {
+    Py_DECREF(pyqos_profile);
+    Py_DECREF(pyqos_depth);
+    return NULL;
+  }
+
+  PyObject * pyqos_reliability = PyLong_FromUnsignedLong(qos_profile->reliability);
+  if (!pyqos_reliability) {
+    Py_DECREF(pyqos_profile);
+    Py_DECREF(pyqos_depth);
+    Py_DECREF(pyqos_history);
+    return NULL;
+  }
+
+  PyObject * pyqos_durability = PyLong_FromUnsignedLong(qos_profile->durability);
+  if (!pyqos_durability) {
+    Py_DECREF(pyqos_profile);
+    Py_DECREF(pyqos_depth);
+    Py_DECREF(pyqos_history);
+    Py_DECREF(pyqos_reliability);
+    return NULL;
+  }
+
+  PyObject * pyqos_avoid_ros_namespace_conventions =
+    PyBool_FromLong(qos_profile->avoid_ros_namespace_conventions);
+  if (!pyqos_avoid_ros_namespace_conventions) {
+    Py_DECREF(pyqos_profile);
+    Py_DECREF(pyqos_depth);
+    Py_DECREF(pyqos_history);
+    Py_DECREF(pyqos_reliability);
+    Py_DECREF(pyqos_durability);
+    return NULL;
+  }
+
+  // A success returns 0, and a failure returns -1
+  int set_result = 0;
+  set_result += PyObject_SetAttrString(pyqos_profile, "depth", pyqos_depth);
+  set_result += PyObject_SetAttrString(pyqos_profile, "history", pyqos_history);
+  set_result += PyObject_SetAttrString(pyqos_profile, "reliability", pyqos_reliability);
+  set_result += PyObject_SetAttrString(pyqos_profile, "durability", pyqos_durability);
+  set_result += PyObject_SetAttrString(pyqos_profile,
+      "avoid_ros_namespace_conventions", pyqos_avoid_ros_namespace_conventions);
+
+  if (0 != set_result) {
+    Py_DECREF(pyqos_depth);
```

Comment:
I think these must always be `DECREF`'d after calling `PyObject_SetAttrString` because it [`INCREF`'s internally](https://github.com/python/cpython/blob/bb86bf4c4eaa30b1f5192dab9f389ce0bb61114d/Objects/dictobject.c#L1029) rather than stealing a reference.

Reply:
Fixed 52afeca